File: Dynamic-Programming/longest-increasing-subsequence.c
/*
Given an unsorted array of integers, find the length of longest increasing subsequence.
Example:
Input: [10,9,2,5,3,7,101,18]
Output: 4 
Explanation: The longest increasing subsequence is [2,3,7,101], therefore the length is 4. 
Note:
	• There may be more than one LIS combination, it is only necessary for you to return the length.
	• Your algorithm should run in O(n2) complexity.

Youtube - https://www.youtube.com/watch?v=CE2b_-XfVDk
*/


int max(int a, int b) {
    if (a > b) return a;
    return b;
}

int lengthOfLIS(int* nums, int numsSize) {
    if (nums == NULL || numsSize == 0) {
        return 0;
    }
    int lis[numsSize];
    int i, j, len=-1;
    
    // set the longest subsequence to at least 1 for each number
    for (i=0; i<numsSize; i++) {
        lis[i] = 1;
    }
    
    for (i=0; i<numsSize; i++) {
        j=0;
        while (j < i) {
            if (nums[j] < nums[i]) {
                lis[i] = max(lis[i], lis[j]+1);
            }
            j++;
        }
    }
    for (i=0; i<numsSize; i++) {
        len = max(len, lis[i]);
    }
    return len;
}

File: Dynamic-Programming/java/LongestIncreasingSubsequence.java
/**
Given an integer array nums, return the length of the longest strictly increasing subsequence.

A subsequence is a sequence that can be derived from an array by deleting some or no elements without changing the order of the remaining elements. For example, [3,6,2,7] is a subsequence of the array [0,3,1,6,2,2,7].

 

Example 1:

Input: nums = [10,9,2,5,3,7,101,18]
Output: 4
Explanation: The longest increasing subsequence is [2,3,7,101], therefore the length is 4.
Example 2:

Input: nums = [0,1,0,3,2,3]
Output: 4
Example 3:

Input: nums = [7,7,7,7,7,7,7]
Output: 1

@Leetcode - https://leetcode.com/problems/longest-increasing-subsequence/
@Youtube - https://www.youtube.com/watch?v=CE2b_-XfVDk
 */
class Solution {
    public int lengthOfLIS(int[] nums) {
        if (nums == null || nums.length == 0) {
            return 0;
        }
        int[] dp = new int[nums.length];
        int result = 1;
        
        // initialize all the elements to 1
        for (int i=0; i<nums.length; i++) {
            dp[i] = 1;
        }
        
        for (int i=1; i<nums.length; i++) {
            for (int j=0; j<i; j++) {
                if (nums[j] < nums[i]) {
                    dp[i] = Math.max(dp[i], dp[j]+1);
                    result = Math.max(result, dp[i]);
                }
            }
        }
        return result;
    }
}
